### PR TITLE
[SPARK-49119][SQL] Fix the inconsistency of syntax `show columns` between v1 and v2

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to 
- fix the `inconsistency` of syntax `show columns` between `v1` and `v2`.
- assign a name `SHOW_COLUMNS_WITH_CONFLICT_NAMESPACE` to the error condition `_LEGACY_ERROR_TEMP_1057`.
- unify v1 and v2 `SHOW COLUMNS ...` tests.
- move some UT related to `SHOW COLUMNS` from `DDLSuite` to `command/ShowColumnsSuiteBase` or `v1/ShowColumnsSuiteBase`.
- move some UT related to `SHOW COLUMNS` from `DDLParserSuite` and `ErrorParserSuite` to `ShowColumnsParserSuite`.

### Why are the changes needed?
In `AstBuilder`, we have `a comment` that explains as follows:
https://github.com/apache/spark/blob/2a752105091ef95f994526b15bae2159657c8ed0/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala#L5054-L5055
However, in our v2 of the syntax `show columns` implementation, we `did not` perform the above checks, as shown below:

```
  withNamespaceAndTable("ns", "tbl") { t =>
      sql(s"CREATE TABLE $t (col1 int, col2 string) $defaultUsing")
      sql(s"SHOW COLUMNS IN $t IN ns1")
    }
```

- Before (inconsistent, v1 will fail, but v2 will success)
  v1:
  ```
  [SHOW_COLUMNS_WITH_CONFLICT_NAMESPACE] SHOW COLUMNS with conflicting namespace: `ns1` != `ns`.
  ```

  v2:
  ```
  Execute successfully.
  ```

#### so, we should fix it.

- After (consistent, v1 & v2 all will fail)
  v1:
  ```
  [SHOW_COLUMNS_WITH_CONFLICT_NAMESPACE] SHOW COLUMNS with conflicting namespace: `ns1` != `ns`.
  ```

  v2:
  ```
  [SHOW_COLUMNS_WITH_CONFLICT_NAMESPACE] SHOW COLUMNS with conflicting namespace: `ns1` != `ns`.
  ```

### Does this PR introduce _any_ user-facing change?
Yes, for v2 tables, in syntax `SHOW COLUMNS {FROM | IN} {tableName} {FROM | IN} {namespace}`, if the namespace (`second parameter`) is different from the namespace of the table(`first parameter`), the command will succeed without any awareness before this PR, after this PR, it will report an error.

### How was this patch tested?
Add new UT.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 1]
```diff
@@ -1385,8 +1385,10 @@ abstract class DDLSuite extends QueryTest with DDLSuiteBase {
             exception = intercept[AnalysisException] {
               sql(s"SHOW COLUMNS IN $db.showcolumn FROM ${db.toUpperCase(Locale.ROOT)}")
             },
-            errorClass = "_LEGACY_ERROR_TEMP_1057",
-            parameters = Map("dbA" -> db.toUpperCase(Locale.ROOT), "dbB" -> db)
+            errorClass = "SHOW_COLUMNS_WITH_CONFLICT_NAMESPACE",
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:1388]: shall we move all the related test cases to the new `ShowColumnsSuiteBase`?

Comment (panbingkun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:1388]: Okay, let me to do it.

Comment (panbingkun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: > shall we move all the related test cases to the new `ShowColumnsSuiteBase`?

Done.

**Comment by yaooqinn** [comment]
In my opinion, when constructing the AST, we should always prepend the "ns" identifier for the table identifier. This is because the syntax states that "a.b.c" is IN/FROM "x.y.z". I can't help but think of concatenating them as "x.y.z.a.b.c" instead of zipping. Currently, we only do this for tables without a specified namespace field.

**Comment by cloud-fan** [comment]
In my opinion, we should not support `SHOW COLUMNS t FROM ...` in the first place. A fully qualified table name should do the work. But the ship is already sailed and this is what it is now.

**Review by cloud-fan — APPROVED** [review]

**Comment by yaooqinn** [comment]
> In my opinion, we should not support `SHOW COLUMNS t FROM ...` in the first place. A fully qualified table name should do the work, there is no need for the FROM clause. But the ship is already sailed and this is what it is now.

FROM and IN are interchangeable. Identifying the tbl or ns is based on the order and the 2nd FROM and IN  clause is identified as ns

**Comment by cloud-fan** [comment]
You can identify a table with the qualified name. This is a common syntax in many other commands. `SHOW COLUMNS FROM t FROM schema` has no benefit but welcomes confusion.

**Comment by yaooqinn** [comment]
So, I recommend prepending it in the first step for the relation resolution, which can somewhat decrease the confusion

**Comment by panbingkun** [comment]
- Yes, in the syntax `SHOW COLUMNS ... {FROM | IN} ... {FROM | IN} ...`, the second clause `{FROM | IN} ..` is actually unnecessary, but due to historical reasons, if we change it to ``SHOW COLUMNS ... {FROM | IN} ...`, it is a `break change` (at least when the end-user run it, it will generate an error). 
- The modification here is only to align this syntax in v1 and v2, `not checking` can cause confusion for end-users

**Comment by cloud-fan** [comment]
> ... which can somewhat decrease the confusion

Yes, but I don't think we should make a breaking change only to reduce confusion for a legacy/unrecommanded feature.

**Comment by yaooqinn** [comment]
> make a breaking change

I acknowledge that I will cause a breaking change for a corner case for V1 tables, `SHOW COLUMNS IN DB.TBL IN DB`.

**Comment by cloud-fan** [comment]
@panbingkun can you update `Does this PR introduce any user-facing change`? I think it does affect v2 tables.

**Comment by panbingkun** [comment]
> @panbingkun can you update `Does this PR introduce any user-facing change`? I think it does affect v2 tables.

Okay

**Comment by panbingkun** [comment]
> @panbingkun can you update `Does this PR introduce any user-facing change`? I think it does affect v2 tables.

Updated.

**Comment by cloud-fan** [comment]
since it introduces a new error, shall we add a legacy config?

**Comment by panbingkun** [comment]
> since it introduces a new error, shall we add a legacy config?

Updated. (`spark.sql.legacy.showColumnsCheckNamespace`)

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowColumnsSuiteBase.scala [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+          sql(s"SHOW COLUMNS IN tbl IN ns1")
+        },
+        errorClass = "TABLE_OR_VIEW_NOT_FOUND",
+        parameters = Map("relationName" -> "`ns1`.`tbl`"),
+        context = ExpectedContext(fragment = "tbl", start = 16, stop = 18)
+      )
+    }
+  }
+
+  test("the namespace of the table conflicts with the specified namespace") {
```
Comment (panbingkun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowColumnsSuiteBase.scala:68]: Refactored the UT and perform testing on the configuration `spark.sql.legacy.showColumnsCheckNamespace` here

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -5118,6 +5118,14 @@ object SQLConf {
     .booleanConf
     .createWithDefault(false)
 
+  val LEGACY_SHOW_COLUMNS_CHECK_NAMESPACE = buildConf("spark.sql.legacy.showColumnsCheckNamespace")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:5121]: usually true means legacy behavior. How about `spark.sql.legacy.showColumnsNoCheck`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowColumnsSuiteBase.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowColumnsSuiteBase.scala:68]: hmm, shouldn't this config only apply to v2 command?

Comment (panbingkun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowColumnsSuiteBase.scala:68]: Yes, I originally planned to unify this testing logic in `command/ShowColumnsSuiteBase` and check if it is v2 as follows:
image

However, from the `perspective` you mentioned, this case seems to have `misled` the developers. 
Let's separate it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 3 above)
Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Updated.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
@yaooqinn you are right! We don't need a config then :)

**Comment by panbingkun** [comment]
> @yaooqinn you are right! We don't need a config then :)

Okay!

File: common/utils/src/main/resources/error/error-conditions.json [hunk 4]
```diff
@@ -3831,6 +3831,12 @@
     ],
     "sqlState" : "42K08"
   },
+  "SHOW_COLUMNS_WITH_CONFLICT_NAMESPACE" : {
```
Comment (yaooqinn) [inline common/utils/src/main/resources/error/error-conditions.json:3834]: I'm not whether it's a good practice or not to create a class/category name that nearly matches the detailed content it holds

Comment (panbingkun) [reply common/utils/src/main/resources/error/error-conditions.json:3834]: In the file `error-conditions.json`, there is a similar one
https://github.com/apache/spark/blob/ac5aa7dcb0a5a85cf05a334912305abeda11f6c2/common/utils/src/main/resources/error/error-conditions.json#L704-L709

Comment (yaooqinn) [reply common/utils/src/main/resources/error/error-conditions.json:3834]: That's bad. Users almost read the same content twice in a single error.

**Review by yaooqinn — APPROVED** [review]

File: common/utils/src/main/resources/error/error-conditions.json [hunk 5]
```diff
@@ -3831,6 +3831,12 @@
...
+    "message" : [
+      "SHOW COLUMNS with conflicting namespace: <namespaceA> != <namespaceB>."
```
Comment (yaooqinn) [inline common/utils/src/main/resources/error/error-conditions.json:3836]:
```suggestion
      "SHOW COLUMNS with conflicting namespaces: <namespaceA> != <namespaceB>."
```

**Comment by LuciferYang** [comment]
Merged into master. Thanks @panbingkun @cloud-fan and @yaooqinn

**Comment by panbingkun** [comment]
> Merged into master. Thanks @panbingkun @cloud-fan and @yaooqinn

Thanks all @LuciferYang @cloud-fan @yaooqinn

**Pull request closed without merging**